packages/web/src/pages/asset-detail-page/components/AssetExternalWallets.tsx

```diff
@@ -0,0 +1,264 @@
+import { useCallback, useContext, useMemo, useState } from 'react'
+
+import {
+  useRemoveConnectedWallet,
+  QUERY_KEYS,
+  useCurrentUserId
+} from '@audius/common/api'
+import { Chain } from '@audius/common/models'
+import { useUserCoin } from '@audius/common/src/api/tan-query/coins/useUserCoin'
+import { shortenSPLAddress } from '@audius/common/utils'
+import {
+  Button,
+  Flex,
+  IconCopy,
+  IconLogoCircleSOL,
+  IconTrash,
+  IconButton,
+  IconKebabHorizontal,
+  Paper,
+  PopupMenu,
+  PopupMenuItem,
+  Text,
+  Box,
+  Skeleton,
+  Divider,
+  IconLogoCircle
+} from '@audius/harmony'
+import { UserCoinAccount } from '@audius/sdk'
+import { useQueryClient } from '@tanstack/react-query'
+
+import { ToastContext } from 'components/toast/ToastContext'
+import { copyToClipboard } from 'utils/clipboardUtil'
+
+import {
+  AlreadyAssociatedError,
+  useConnectAndAssociateWallets
+} from '../../../hooks/useConnectAndAssociateWallets'
+import { AssetDetailProps } from '../types'
+
+const COPIED_TOAST_TIMEOUT = 2000
+
+const messages = {
+  noBalanceTitle: 'Link External Wallet',
+  hasBalanceTitle: 'Balance Breakdown',
+  description:
+    'Link an external wallet to take advantage of in-app features, and take full control of your assets.',
+  loadingText: 'Loading...',
+  buttonText: 'Add External Wallet',
+  manageWallets: 'Manage External Wallets',
+  copied: 'Copied To Clipboard!',
+  audio: '$AUDIO',
+  copy: 'Copy Wallet Address',
+  remove: 'Remove Wallet',
+  options: 'Options',
+  newWalletConnected: 'New Wallet Successfully Connected!',
+  walletRemoved: 'Wallet Successfully Removed!',
+  error: 'Something went wrong. Please try again.',
+  walletAlreadyAdded: 'No new wallets selected to connect.'
+}
+
+type WalletRowProps = {
+  mint: string
+  decimals: number
+} & UserCoinAccount
+
+const WalletRow = ({
+  owner: address,
+  balance,
+  isInAppWallet,
+  decimals
+}: WalletRowProps) => {
+  const { toast } = useContext(ToastContext)
+  const [isRemoving, setIsRemoving] = useState(false)
+  const queryClient = useQueryClient()
+  const { data: currentUserId } = useCurrentUserId()
+  const copyAddressToClipboard = useCallback(() => {
+    copyToClipboard(address)
+    toast(messages.copied, COPIED_TOAST_TIMEOUT)
+  }, [address, toast])
+
+  const { mutateAsync: removeConnectedWalletAsync } = useRemoveConnectedWallet()
+
+  const handleRemove = useCallback(async () => {
+    try {
+      setIsRemoving(true)
+      await removeConnectedWalletAsync({
+        wallet: { address, chain: Chain.Sol }
+      })
+      await queryClient.invalidateQueries({
+        queryKey: [QUERY_KEYS.userCoin, currentUserId]
+      })
+      toast('Wallet removed successfully!')
+    } catch (e) {
+      toast('Error removing wallet')
+    } finally {
+      setIsRemoving(false)
+    }
+  }, [removeConnectedWalletAsync, address, queryClient, toast, currentUserId])
+
+  const items: PopupMenuItem[] = useMemo(
```

Comment:
nit: think this is easier to read with a normal conditional and a `.push` if the `isInAppWallet` condition is true